intel_extension_for_transformers/transformers/modeling/kv_cahe_compression/h2o_sim_drop/modify_gptneox.py

```diff
@@ -0,0 +1,278 @@
+    accumulated_attention_score[:,:,heavy_budget+padding_length:] = 0
+    if padding_length > 0:
+        accumulated_attention_score[:,:,:padding_length] = 0
+
+    mask_bottom = torch.zeros_like(attn_weights, dtype=torch.bool)
+    mask_bottom[:,:, padding_length:heavy_budget+padding_length, padding_length:heavy_budget+padding_length] = True
+
+    for token_index in range(heavy_budget+padding_length, seq_length):
+
+        tmp_attn_index = nn.functional.softmax(attn_weights[:,:,token_index,:], dim=-1, dtype=torch.float32).to(dtype_attn_weights)
+        _, tmp_topk_index = accumulated_attention_score.topk(k=heavy_budget-1, dim=-1)
+        zeros_index = torch.zeros_like(tmp_attn_index, dtype=torch.bool)
+        mask_bottom_index = zeros_index.scatter(-1, tmp_topk_index, True) #(head, keys)
+        mask_bottom_index[:,:, token_index] = True
+
+        mask_bottom[:,:,token_index,:] = mask_bottom_index
+        accumulated_attention_score += tmp_attn_index
+        accumulated_attention_score = accumulated_attention_score * mask_bottom_index
+
+    return mask_bottom
+
+def sanity_check(mask):
+    # mask (head, query, key)
+    ones = torch.ones_like(mask)
+    ones = torch.triu(ones, diagonal=0)
+    mask_bottom = torch.logical_or(mask, ones)
+
+    error_cnt = 0
+    for i in range(mask_bottom.shape[1]-1):
+        index = mask_bottom[:,i,:].eq(0).unsqueeze(1)
+        index[:,i:]=0
+        error_cnt += (mask_bottom[:,i:,:] * index).sum().item()
+    print(error_cnt)
+    return error_cnt
+
+class GPTNeoXAttention_Mask(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.num_attention_heads = config.num_attention_heads
+        self.hidden_size = config.hidden_size
+        self.head_size = self.hidden_size // self.num_attention_heads
+        self.rotary_ndims = int(self.head_size * config.rotary_pct)
+        max_positions = config.max_position_embeddings
+        self.register_buffer(
+            "bias",
+            torch.tril(torch.ones((max_positions, max_positions), dtype=torch.bool)).view(
+                1, 1, max_positions, max_positions
+            ),
+        )
+        self.register_buffer("masked_bias", torch.tensor(-1e9))
+        # self.rotary_emb = RotaryEmbedding(
+        self.rotary_emb = GPTNeoXRotaryEmbedding(
+            self.rotary_ndims, config.max_position_embeddings, base=config.rotary_emb_base
+        )
+        self.register_buffer(
+            "norm_factor",
+            torch.sqrt(torch.tensor(self.head_size, dtype=torch.float32)).to(torch.get_default_dtype()),
+            persistent=False,
+        )
+        self.query_key_value = nn.Linear(config.hidden_size, 3 * config.hidden_size)
+        self.dense = nn.Linear(config.hidden_size, config.hidden_size)
+
+        self.heavy_budget_ratio = config.heavy_ratio
+        self.recent_budget_ratio = config.recent_ratio
```

Comment:
for all the kv pruning methods, let's add one more hyperparameters, seqlen_threshold, if the kept seqlen is less than this threshold, no pruning is conducted.